src/libs/print_settings.c

```diff
@@ -280,101 +271,187 @@ _print_button_clicked (GtkWidget *widget, gpointer user_data)
     fprintf(stderr, "failed to create temporary pdf for printing\n");
-    return;
+    return 1;
   }
   close(fd);
 
   const int icc_id = 0;
 
-  dt_pdf_t *pdf = dt_pdf_start(filename, page_width, page_height, ps->prt.printer.resolution, DT_PDF_STREAM_ENCODER_FLATE);
+  dt_pdf_t *pdf = dt_pdf_start(params->pdf_filename, page_width, page_height, params->prt.printer.resolution, DT_PDF_STREAM_ENCODER_FLATE);
 
 /*
   // ??? should a profile be embedded here?
   if (*printer_profile)
     icc_id = dt_pdf_add_icc(pdf, printer_profile);
 */
-  dt_pdf_image_t *pdf_image = dt_pdf_add_image(pdf, (uint8_t *)dat.ps->buf, dat.width, dat.height, 8, icc_id, 0.0);
+  params->pdf_image = dt_pdf_add_image(pdf, (uint8_t *)params->buf, dat.width, dat.height, 8, icc_id, 0.0);
 
   //  PDF bounding-box has origin on bottom-left
-  pdf_image->bb_x      = dt_pdf_pixel_to_point((float)margin_left, ps->prt.printer.resolution);
-  pdf_image->bb_y      = dt_pdf_pixel_to_point((float)margin_bottom, ps->prt.printer.resolution);
-  pdf_image->bb_width  = dt_pdf_pixel_to_point((float)iwidth, ps->prt.printer.resolution);
-  pdf_image->bb_height = dt_pdf_pixel_to_point((float)iheight, ps->prt.printer.resolution);
+  params->pdf_image->bb_x      = dt_pdf_pixel_to_point((float)margin_left, params->prt.printer.resolution);
+  params->pdf_image->bb_y      = dt_pdf_pixel_to_point((float)margin_bottom, params->prt.printer.resolution);
+  params->pdf_image->bb_width  = dt_pdf_pixel_to_point((float)iwidth, params->prt.printer.resolution);
+  params->pdf_image->bb_height = dt_pdf_pixel_to_point((float)iheight, params->prt.printer.resolution);
 
-  if (ps->prt.page.landscape && (dat.width > dat.height))
-    pdf_image->rotate_to_fit = TRUE;
+  if (params->prt.page.landscape && (dat.width > dat.height))
+    params->pdf_image->rotate_to_fit = TRUE;
   else
-    pdf_image->rotate_to_fit = FALSE;
-
-  dt_pdf_page_t *pdf_page = dt_pdf_add_page(pdf, &pdf_image, 1);
-  dt_pdf_finish(pdf, &pdf_page, 1);
+    params->pdf_image->rotate_to_fit = FALSE;
 
-  // free memory
+  params->pdf_page = dt_pdf_add_page(pdf, &params->pdf_image, 1);
+  dt_pdf_finish(pdf, &params->pdf_page, 1);
 
-  free (dat.ps->buf);
-  free (pdf_image);
-  free (pdf_page);
+  if(dt_control_job_get_state(job) == DT_JOB_STATE_CANCELLED) return 0;
+  dt_control_job_set_progress(job, 0.95);
 
   // send to CUPS
 
-  dt_print_file (imgid, filename, &ps->prt);
-
-  g_unlink(filename);
+  dt_print_file (params->imgid, params->pdf_filename, params->job_title, &params->prt);
+  dt_control_job_set_progress(job, 1.0);
 
   // add tag for this image
 
   char tag[256] = { 0 };
   guint tagid = 0;
-  snprintf (tag, sizeof(tag), "darktable|printed|%s", ps->prt.printer.name);
+  snprintf (tag, sizeof(tag), "darktable|printed|%s", params->prt.printer.name);
   dt_tag_new(tag, &tagid);
-  dt_tag_attach(tagid, imgid);
+  dt_tag_attach(tagid, params->imgid);
+
+  return 0;
+}
+
+static void _print_job_cleanup(void *p)
```

Comment:
I would pass directly a "dt_lib_print_job_t **p" and then ends with free(*p); *p=NULL;

To ensure that the passed parameter do not keep an address to a released memory chunk.

Reply:
I see that this is a callback passed to the job control. Not possible I suppose then.

Reply:
It looks like `dt_control_job_dispose()` keeps the last pointer to the job params, then it `free()`'s the job structure which includes the pointer. https://github.com/darktable-org/darktable/blob/afe30d62f550a6243e54088c2b8cf97e6fee62c3/src/control/jobs.c#L142-L152

The extant job code already deals with getting rid of the last reference to the released memory?

Reply:
No, job->params is not released in dt_control_job_dispose(). So you code is correct and we don't have a dangling reference indeed as the top-level struct (_dt_jot_t) is freed. All good I suppose.